hs/app/reach/Main.hs

```diff
@@ -65,15 +65,161 @@ writeFrom :: FilePath -> App
 writeFrom p = asks e_dirEmbed >>= liftIO . T.readFile . (</> p) >>= write
 
 
-runSubScript :: App -> Env -> IO T.Text
-runSubScript a e = readIORef (e_effect e) >>= \case
+_runSubScript :: App -> Env -> IO T.Text
+_runSubScript a e = readIORef (e_effect e) >>= \case
   Script _ -> do
     r <- newIORef $ Script ""
     Script r' <- runReaderT a (e { e_effect = r }) >> readIORef r
     pure r'
   _ -> U.impossible "`runSubScript` may only be applied to a `Script`"
 
 
+--------------------------------------------------------------------------------
+serviceConnector :: FilePath -> Env -> T.Text -> [T.Text] -> IO T.Text
+serviceConnector name Env {..} rv ports = do
+  let indentPorts = "  "
+  let ports' = case ports of
+        [] -> indentPorts <> "[]"
+        ps -> T.unlines $ map ((indentPorts <> "- ") <>) ps
+
+  fmt <- T.readFile $ e_dirEmbed
+    </> "sh" </> "_common" </> "_docker-compose" </> "service-" <> name <> ".yml"
+
+  pure
+    . swap "REACH_VERSION" rv
+    . swap "PORTS" ports'
+    $ fmt
+
+
+serviceDevnetAlgo :: Env -> T.Text -> [T.Text] -> IO T.Text
+serviceDevnetAlgo env rv ports = do
+  dn <- serviceConnector "devnet-algo" env rv ports
+  db <- serviceConnector "devnet-algo-postgres-db" env rv ports
+  pure $ dn <> db
+
+
+serviceDevnetCfx :: Env -> T.Text -> [T.Text] -> IO T.Text
+serviceDevnetCfx = serviceConnector "devnet-cfx"
+
+
+serviceDevnetEth :: Env -> T.Text -> [T.Text] -> IO T.Text
+serviceDevnetEth = serviceConnector "devnet-eth"
+
+
+-- TODO dynamic `appService'` definition
+mkComposeYml :: T.Text -> T.Text -> [T.Text] -> T.Text
+mkComposeYml appService' appImageTag' svs =
+  let svs' = T.intercalate "" svs
+   in [N.text|
+      version: '3.4'
+      services:
+        $svs'
+
+        $appService':
+          image: $appImageTag'
+          depends_on:
+            - ethereum-devnet
+          environment:
+            - REACH_DEBUG
+            - REACH_CONNECTOR_MODE=ETH-test-dockerized-geth
+            - ETH_NODE_URI=http://ethereum-devnet:8545
+      |]
+
+
+--------------------------------------------------------------------------------
+data ConnectorMode
```

Comment:
I think this type should be `(Connector, Mode)` where `Connector` is `ALGO | CFX | ETH` and `Mode` is `Devnet|Live|Browser`

Reply:
I used `data ConnectorMode = ConnectorMode Connector Mode` because there's already a `Show` instance for tuples that would overlap here.